## Importing Modules

In [12]:
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, KFold

## Reading the file

In [13]:
df = pd.read_csv('../chicago_crime_dataset.csv')

## Preprocessing and filtering the data

In [ ]:
df = df.dropna()
df['Date'] = pd.to_datetime(df['Date'])
df['timestamp'] = df['Date'].dt.strftime('%H:%M')
df = df[df['timestamp'] != '00:00']

## Splitting Features into Given and Target Features

In [15]:
final_features = [ 'Primary Type', 'Location Description', 'Domestic',  'District',  'Year','timestamp','Arrest']

df[final_features].head(10)
df[final_features].info()

<class 'pandas.DataFrame'>
Index: 114690 entries, 0 to 120758
Data columns (total 7 columns):
 #   Column                Non-Null Count   Dtype
---  ------                --------------   -----
 0   Primary Type          114690 non-null  str  
 1   Location Description  114690 non-null  str  
 2   Domestic              114690 non-null  bool 
 3   District              114690 non-null  int64
 4   Year                  114690 non-null  int64
 5   timestamp             114690 non-null  str  
 6   Arrest                114690 non-null  bool 
dtypes: bool(2), int64(2), str(3)
memory usage: 5.5 MB


In [16]:
X = df[final_features].drop('Arrest', axis=1)
y = df['Arrest']

## Preprocessing and encoding

In [17]:
preprocessor = ColumnTransformer([
    ('onehot', OneHotEncoder(handle_unknown='ignore'), ['Primary Type', 'Location Description', 'Domestic', 'timestamp'])
], remainder='passthrough')

## Model Initialisation

In [18]:
model1 = RandomForestClassifier()
model2 = DecisionTreeClassifier()

## Pipeline

In [19]:
pipeline1 = Pipeline([('preprocessor', preprocessor), ('classifier', model1)])
pipeline2 = Pipeline([('preprocessor', preprocessor), ('classifier', model2)])

## K Fold Cross Validation

In [20]:
scores_model1 = cross_val_score(pipeline1, X, y, cv=5)
scores_model2 = cross_val_score(pipeline2, X, y, cv=5)

## Printing and comparing scores



In [21]:
print("RandomForestClassifier : ", scores_model1.mean())
print("DecisionTreeClassifier : ", scores_model2.mean())

RandomForestClassifier :  0.7986834074461593
DecisionTreeClassifier :  0.7711831894672596


## Final View

DecisionTreeClassifer and RandomForestClassifer models have been used to compare on the given dataset using K-Flip Cross Validation.

As per the observation on the mean scores of both the models, RandomForestClassifer has been more accurate than DecisionTreeCLassifier, with their accuracies being 79.868% and 77.118% respectively
